Track timing of consensus stages

vicsn opened the pull request:
## Motivation

This PR introduces a method to track consensus events for visualization, building on top of @kpandl 's earlier work. My understanding of the image below is that under load, block processing is the bottleneck, not certificate signing.

This "ugly" approach mirrors the way we track `pub static R1CS_HASHES` in snarkVM for certain tests only.

Open TODOs:
- [ ] Add a few logs so we can observe the timing in production

<img width="789" height="944" alt="Screenshot 2025-10-12 at 21 31 33" src="https://github.com/user-attachments/assets/25e3563b-c64c-47d4-a867-3d354e0b9106" />

<img width="792" height="925" alt="Screenshot 2025-10-13 at 20 23 52" src="https://github.com/user-attachments/assets/15cf3aed-6f00-41fe-bf1a-8c5f91557da6" />

vicsn committed:
Track timing of consensus stages

ljedrz commented on node/bft/src/helpers/timing.rs:
```diff
@@ -0,0 +1,317 @@
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct TimingEvent {
+    pub round: u64,
+    pub timestamp: SystemTime,
```
`SystemTime` is not monotonic; it would be more reliable to instead use the `Instant`, especially for the shorter events.

vicsn replied:
When trying to replace it, compiler complains that it can't derive Serialize/Deserialize on Instant.

Is this an issue in practice? The created graph seems quite faithful and we don't care about even a millisecond difference at these scales.

ljedrz replied:
hmm, as long as we don't perform uncheck subtraction between such values and we don't worry about tiny events, it should be fine

ljedrz commented on node/bft/src/helpers/timing.rs:
```diff
@@ -0,0 +1,317 @@
+pub struct TimingEvent {
+    pub round: u64,
+    pub timestamp: SystemTime,
+    pub event_type: String,
```
```suggestion
    pub event_type: &'static str,
```
judging by the current use

vicsn replied:
Can't auto-deserialize into a static reference:

```
error: lifetime may not live long enough
  --> node/bft/src/helpers/timing.rs:62:5
   |
59 | #[derive(Debug, Clone, Serialize, Deserialize)]
   |                                   ----------- lifetime `'de` defined here
...
62 |     pub proposal_seen: Vec<TimingEvent>,
   |     ^^^ requires that `'de` must outlive `'static`
```

ljedrz commented on node/bft/src/helpers/timing.rs:
```diff
@@ -0,0 +1,317 @@
+use serde::{Deserialize, Serialize};
+
+/// Global storage for round-based event data
+static ROUND_EVENTS: Lazy<Arc<RwLock<HashMap<u64, RoundEvents>>>> = Lazy::new(|| Arc::new(RwLock::new(HashMap::new())));
```
```suggestion
static ROUND_EVENTS: Lazy<Arc<Mutex<HashMap<u64, RoundEvents>>>> = Lazy::new(|| Arc::new(RwLock::new(HashMap::new())));
```
I'm assuming we're only interested in the final values.

vicsn replied:
Leaving as is for now as a bandaid for the fact that we inefficiently read out on every block.

ljedrz commented on node/bft/src/helpers/timing.rs:
```diff
@@ -0,0 +1,317 @@
+static ROUND_EVENTS: Lazy<Arc<RwLock<HashMap<u64, RoundEvents>>>> = Lazy::new(|| Arc::new(RwLock::new(HashMap::new())));
+
+/// Global storage for subdag-based timing data (using lowest/highest rounds from subdag)
+static SUBDAG_TIMINGS: Lazy<Arc<RwLock<HashMap<(u64, u64), SubdagTimings>>>> =
```
```suggestion
static SUBDAG_TIMINGS: Lazy<Arc<Mutex<HashMap<(u64, u64), SubdagTimings>>>> =
```
ditto: as long as we're not going to be actively inspecting these values with multiple readers, a `Mutex` will suffice

ljedrz commented on node/bft/src/helpers/timing.rs:
```diff
@@ -0,0 +1,317 @@
+                    self.advance_to_next_block = Some((start, Some(now)));
+                }
+            }
+        }
```
should we notify the user if `None` was found in any of those (indicating that a `start_stage` may have been omitted by accident)?

vicsn replied:
Good idea: d9492d025

ljedrz commented on node/bft/src/helpers/timing.rs:
```diff
@@ -0,0 +1,317 @@
+
+/// Clean up old timing entries to prevent memory growth
+/// Keeps only the most recent `keep_count` entries for each type
+pub fn cleanup_old_entries(keep_count: usize) {
```
Considering this use case, it would be more performant to use a `VecDeque` and use the round as the index (offset by the 1st round in the collection, which would be stored separately), but this should be good enough for ad-hoc dev applications.

vicsn replied:
Removed the function f462afde7

ljedrz commented:
> dumping all of the data into a global lazy static, and only enabling this in dev/test_network mode. Would a pub static also be fine? Or something else?

For dev/test purposes it's perfectly fine; it is only in production where it's an anti-pattern, and should instead be incorporated into the workflow. The alternative would be to store it in the related objects, which would also work, though could be a bit trickier to implement.

> exporting the data to a JSON file, also only in dev/test_network mode.

As long as we don't expect very long measurements/outputs, this is fine.

Oh, and naturally, this still needs to be feature-gated.

vicsn committed:
Guard consensus tracking by test_network feature, remove dead code, use parking_lot lock

vicsn reviewed (commented):
TODO: also measure a stage for `commit_leader_certificate` onwards

vicsn committed:
Add SubdagProcessing, remove is_local

vicsn committed:
Add useful warnings

vicsn committed:
Merge remote-tracking branch 'origin/staging' into consensus_tracking

kaimast commented:
Would it not be better to add a separate feature for this? Or do you think the performance overhead for test networks that do not require timing information is minimal?

vicsn committed:
Introduce test_consensus_tracking feature flag

Using test_network for consensus_tracking could be inefficient for
long-running dev networks.

Merging it with production metrics would take a lot more work.

vicsn commented:
> Would it not be better to add a separate feature for this? Or do you think the performance overhead for test networks that do not require timing information is minimal?

Agreed, made a new `test_consensus_tracking` feature. Specifically for long-running networks, these hacky JSON files might get huge.

vicsn committed:
Merge remote-tracking branch 'origin/staging' into consensus_tracking

vicsn committed:
Update Cargo.lock

ljedrz commented on node/bft/Cargo.toml:
```diff
@@ -44,6 +44,7 @@ test = [
   "snarkos-node-bft-ledger-service/test",
   "snarkos-node-bft-storage-service/test"
 ]
+test_consensus_tracking = [ ]
```
nit: this should ideally include the new dependencies

vicsn replied:
Hmm that's a very tight coupling, I'll leave it out for now

ljedrz commented on node/consensus/src/lib.rs:
```diff
@@ -524,12 +524,73 @@ impl<N: Network> Consensus<N> {
+        // Extract lowest and highest rounds from the subdag
+        let rounds: Vec<u64> = subdag.keys().copied().collect();
+        let _lowest_round = rounds.iter().min().copied().unwrap_or(0);
+        let _highest_round = rounds.iter().max().copied().unwrap_or(0);
```
```suggestion
        let _lowest_round = rounds.first_key_value().map(|(k, _)| *k).unwrap_or(0);
        let _highest_round = rounds.last_key_value().map(|(k, _)| *k).unwrap_or(0);
```

vicsn replied:
0f89bbcd8

ljedrz commented on node/bft/src/helpers/timing.rs:
```diff
@@ -0,0 +1,300 @@
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct TimingSnapshot {
+    pub timestamp: SystemTime,
+    pub round_events: HashMap<String, RoundEvents>,
+    pub subdag_timings: HashMap<String, SubdagTimings>,
```
not sure it's worth it to lose the ordering just to have faster queries only in tests

vicsn replied:
Irrelevant at this stage

ljedrz commented on node/bft/src/helpers/timing.rs:
```diff
@@ -0,0 +1,300 @@
+use serde::{Deserialize, Serialize};
+
+/// Global storage for round-based event data
+static ROUND_EVENTS: Lazy<Arc<RwLock<HashMap<u64, RoundEvents>>>> = Lazy::new(|| Arc::new(RwLock::new(HashMap::new())));
```
```suggestion
static ROUND_EVENTS: Lazy<Arc<RwLock<HashMap<u64, RoundEvents>>>> = Lazy::new(|| Default::default());
```
nit

kaimast replied:
You might also be able to use [`LazyLock`](https://doc.rust-lang.org/std/sync/struct.LazyLock.html) in `std` instead of relying on `OnceCell`.

vicsn replied:
Why is that preferred?

vicsn replied:
0f89bbcd8

ljedrz commented on node/bft/src/helpers/timing.rs:
```diff
@@ -0,0 +1,300 @@
+
+/// Global storage for subdag-based timing data (using lowest/highest rounds from subdag)
+static SUBDAG_TIMINGS: Lazy<Arc<RwLock<HashMap<(u64, u64), SubdagTimings>>>> =
+    Lazy::new(|| Arc::new(RwLock::new(HashMap::new())));
```
```suggestion
    Lazy::new(|| Default::default());
```
same (as long as the complier doesn't find a reason to complain)

vicsn replied:
0f89bbcd8

ljedrz reviewed (commented):
Left a few comments.

kaimast commented on node/consensus/src/lib.rs:
```diff
@@ -524,12 +524,73 @@ impl<N: Network> Consensus<N> {
+        // Export timing data to JSON after block generation
+        #[cfg(feature = "test_consensus_tracking")]
+        {
+            let dev_index = self.bft().primary().gateway().dev().unwrap_or_default();
```
Should this be `0` when not in dev mode? We could alternatively append nothing or `_nodev`.

vicsn replied:
Good catch: 5da19d5b2

kaimast commented on Cargo.toml:
```diff
@@ -281,6 +281,7 @@ serial = [ "snarkos-cli/serial" ]
 test_targets = [ "snarkos-cli/test_targets" ]
 test_consensus_heights = [ "snarkos-cli/test_consensus_heights" ]
 test_network = [ "snarkos-cli/test_network" ]
+test_consensus_tracking = [ "snarkos-node/test_consensus_tracking" ]
```
It would be good to start documenting these new features. Either here or in the README similar to what I did in [`this PR`](https://github.com/ProvableHQ/snarkVM/pull/2887/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R48).

vicsn replied:
ae0388458

kaimast commented on node/consensus/src/lib.rs:
```diff
@@ -524,12 +524,73 @@ impl<N: Network> Consensus<N> {
+        let _lowest_round = rounds.iter().min().copied().unwrap_or(0);
+        let _highest_round = rounds.iter().max().copied().unwrap_or(0);
+
+        #[cfg(feature = "test_consensus_tracking")]
```
It might make sense to add a macro for the `start_subdag_stage`/`end_subdag_stage` pair. That macro could also be no-op if the feature flag isn't set, so we do not have to have this many feature gates in the code.

kaimast replied:
(I can push a commit for with this change, if it is too much of a hassle for you)

vicsn replied:
Agreed it was really ugly, made it more readable: 5da19d5b2

I don't like the marginal improvement of a new macro, but maybe once your other tracing logic works and we can decorate functions with it, it could replace this current approach entirely.

vicsn committed:
Fix style nits

vicsn committed:
Document build features

vicsn committed:
Clean up consensus tracking

vicsn committed:
Fix build

vicsn committed:
Merge remote-tracking branch 'origin/staging' into consensus_tracking

vicsn committed:
Merge remote-tracking branch 'origin/staging' into consensus_tracking
